src/responder/pam/pamsrv_passkey.c

```diff
@@ -0,0 +1,935 @@
+                                 &mappings, NULL);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Incorrectly formatted passkey data [%d]: %s\n",
+                                     ret, sss_strerror(ret));
+            goto done;
+        }
+
+        kh_mappings[i] = talloc_strdup(kh_mappings, mappings[0] + strlen(PASSKEY_PREFIX));
+        if (kh_mappings[i] == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "talloc_strdup key handle failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        public_keys[i] = talloc_strdup(public_keys, mappings[1]);
+        if (public_keys[i] == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "talloc_strdup public key failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+    }
+
+    domain_name = talloc_strdup(tmp_ctx, domain);
+    if (domain_name == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_strdup domain failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    _data->domain = talloc_steal(mem_ctx, domain_name);
+    _data->key_handles = talloc_steal(mem_ctx, kh_mappings);
+    _data->public_keys = talloc_steal(mem_ctx, public_keys);
+    _data->num_passkeys = el->num_values;
+
+    ret = EOK;
+done:
+    talloc_free(tmp_ctx);
+
+    return ret;
+}
+
+static void pam_forwarder_passkey_cb(struct tevent_req *req)
+{
+    struct pam_auth_req *preq = tevent_req_callback_data(req,
+                                                         struct pam_auth_req);
+    errno_t ret = EOK;
+    int child_status;
+
+    ret = pam_passkey_auth_recv(req, &child_status);
+    talloc_free(req);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "PAM passkey auth failed [%d]: %s\n",
+                                 ret, sss_strerror(ret));
+        goto done;
+    }
+
+    preq->pd->pam_status = PAM_SUCCESS;
+    pam_reply(preq);
+
+    return;
+
+done:
+    pam_check_user_done(preq, ret);
+}
+
+void pam_passkey_get_user_done(struct tevent_req *req)
+{
+    int ret;
+    struct passkey_ctx *pctx;
+    bool debug_libfido2 = false;
+    char *domain_name = NULL;
+    int timeout;
+    struct cache_req_result *result;
+    struct pk_child_user_data *pk_data = NULL;
+    enum passkey_user_verification verification = PAM_PASSKEY_VERIFICATION_OMIT;
+
+    pctx = tevent_req_callback_data(req, struct passkey_ctx);
+
+    ret = pam_passkey_get_mapping_recv(pctx, req, &result);
+    talloc_zfree(req);
+    if (ret != EOK && ret != ENOENT) {
+        DEBUG(SSSDBG_OP_FAILURE, "cache_req_user_by_name_attrs_recv failed [%d]: %s.\n",
+                                 ret, sss_strerror(ret));
+        goto done;
+    }
+
+    pk_data = talloc_zero(pctx, struct pk_child_user_data);
+    if (!pk_data) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "pk_data == NULL\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    /* Use dns_name for AD/IPA - for LDAP fallback to domain->name */
+    domain_name = result->domain->dns_name;
```

Comment:
covscan complains that there is no result->domain != NULL check. I think it wouldn't hurt to add one...

Reply:
Thanks Alexander, it is fixed now.